stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+#include "core/divemode.h"
+#include "core/divesite.h"
+#include "core/gas.h"
+#include "core/pref.h"
+#include "core/qthelper.h" // for get_depth_unit() et al.
+#include "core/subsurface-time.h"
+#include <limits>
+#include <QLocale>
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+#define SKIP_EMPTY Qt::SkipEmptyParts
+#else
+#define SKIP_EMPTY QString::SkipEmptyParts
+#endif
+
+static const constexpr double NaN = std::numeric_limits<double>::quiet_NaN();
+
+// Typedefs for year / quarter or month binners
+using year_quarter = std::pair<unsigned short, unsigned short>;
+using year_month = std::pair<unsigned short, unsigned short>;
+
+// Small helper template: add an item to an unsorted vector, if its not already there.
+template<typename T>
+static void add_to_vector_unique(std::vector<T> &v, const T &item)
+{
+	if (std::find(v.begin(), v.end(), item) == v.end())
+		v.push_back(item);
+}
+
+// Small helper: make a comma separeted list of a vector of QStrings
+static QString join_strings(const std::vector<QString> &v)
+{
+	QString res;
+	for (const QString &s: v) {
+		if (!res.isEmpty())
+			res += ", ";
+		res += s;
+	}
+	return res;
+}
+
+// Note: usually I dislike functions defined inside class/struct
+// declarations ("Java style"). However, for brevity this is done
+// in this rather template-heavy source file more or less consistently.
+
+// Templates to define invalid values for variables and test for said values.
+// This is used by the binners: returning such a value means "ignore this dive".
+template<typename T> T invalid_value();
+template<> int invalid_value<int>()
+{
+	return std::numeric_limits<int>::max();
+}
+template<> double invalid_value<double>()
+{
+	return std::numeric_limits<double>::quiet_NaN();
+}
+template<> QString invalid_value<QString>()
+{
+	return QString();
+}
+template<> StatsQuartiles invalid_value<StatsQuartiles>()
+{
+	double NaN = std::numeric_limits<double>::quiet_NaN();
+	return { NaN, NaN, NaN, NaN, NaN };
+}
+
+static bool is_invalid_value(int i)
+{
+	return i == std::numeric_limits<int>::max();
+}
+
+static bool is_invalid_value(double d)
+{
+	return std::isnan(d);
+}
+
+static bool is_invalid_value(const QString &s)
+{
+	return s.isEmpty();
+}
+
+// Currently, we don't support invalid dates - should we?
+static bool is_invalid_value(const year_quarter &)
+{
+	return false;
+}
+
+static bool is_invalid_value(const dive_site *d)
+{
+	return !d;
+}
+
+static bool is_invalid_value(const StatsOperationResults &res)
+{
+	return !res.isValid();
+}
+
+struct gas_bin_t {
+	enum class Type {
+		Air,
+		Oxygen,
+		EAN,
+		Trimix
+	} type;
+	int o2, he;
+	static gas_bin_t air() {
+		return { Type::Air, 0, 0 };
+	}
+	static gas_bin_t oxygen() {
+		return { Type::Oxygen, 0, 0 };
+	}
+	static gas_bin_t ean(int o2) {
+		return { Type::EAN, o2, 0 };
+	}
+	static gas_bin_t trimix(int o2, int h2) {
+		return { Type::Trimix, o2, h2 };
+	}
+};
+
+static bool is_invalid_value(const gas_bin_t &)
+{
+	return false;
```

Comment:
not sure if this is permanently prevented somehow, but `o2+he > 1000` or either of the individual values `< 0` or `> 1000` all seem like valid `invalid` criteria

Reply:
Yes - we never bin invalid gases, so invalid gas-bins will never exist:
```
                        struct gasmix mix = d->cylinders.cylinders[i].gasmix;
                        if (gasmix_is_invalid(mix))
                                continue;
```
Added a comment.